Yet more TBDGen work: mostly classes.

huonw opened the pull request:
~~There's some seemingly unfortunate stuff around ObjC classes and metaclasses, which are currently public even for private classes, but that's a bug rather than something blocking this feature.~~ (I think this wasn't as bad as it sounds, and the problem that was misleading me was fixed by "ObjC class aliases are as visible as their aliasee.")

huonw committed:
[TBDGen] Generic types have fewer symbols.

huonw committed:
[TBDGen] `main` is a symbol.

huonw committed:
[TBDGen] Only normal conformances get symbols.

huonw committed:
[AST] Classes know their appropriate metaclass.

huonw committed:
[TBDGen] ObjC classes and metaclasses.

huonw committed:
[TBDGen] ObjC classes don't have public destructors.

huonw committed:
[TBDGen] Resilient classes can have indirect fields.

huonw committed:
[IRGen] ObjC class aliases are as visible as their aliasee.

huonw commented:
@swift-ci Please smoke test

huonw commented:
@swift-ci Please smoke test

huonw commented:
@swift-ci Please smoke test

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -2584,6 +2584,13 @@ ObjCClassKind ClassDecl::checkObjCAncestry() const {
+ClassDecl::MetaclassKind ClassDecl::getMetaclassKind() const {
+  assert(getASTContext().LangOpts.EnableObjCInterop &&
+         "querying metaclass kind without objc interop");
+  auto objc = checkObjCAncestry() != ObjCClassKind::NonObjC || hasClangNode();
```
Is it ever the case that checkObjCAncestry() == NonObjC, but hasClangNode() is true?

huonw replied:
I don't know. Based on the blame it seems some person called @slavapestov was the most recent to touch this (in https://github.com/apple/swift/commit/871d71a555ddeb5a572cc90ea8415b782c710db6 ), but the logic has been here since @rjmccall made the compiler decide between objc and swift metaclasses more than 4 years ago. I'll try removing it.

huonw replied:
Done.

slavapestov commented:
Were you going to fix the visibility? :-)

huonw commented:
If you agree it's a bug, I'll file a bug, but I won't be fixing it until TBD (etc.) are brought up since it isn't blocking.

slavapestov commented:
Depending on how much work it would be I would suggest trying to tackle a few of these small fixes as you go along instead of punting them until later. Remember, the whole point of doing these features is not the feature itself, but the cleanups and fixes they allow us to sneak in ;-)

huonw committed:
[AST] No redundant clang node check for metaclass ObjC-ness.

It seems that any type that has a clang node will have a non-NonObjC
ObjCClassKind, so this Or expression was essentially `foo() || true`.

huonw commented:
@swift-ci Please smoke test and merge

huonw commented:
@swift-ci Please smoke test

huonw commented:
@swift-ci Please smoke test Linux
